src/libraries/System.Text.Json/src/System/Text/Json/Serialization/AsyncEnumerableStreamingDeserializer.cs

```diff
@@ -0,0 +1,113 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Text.Json.Serialization
+{
+    internal sealed class AsyncEnumerableStreamingDeserializer<TValue> : IAsyncEnumerable<TValue>
+    {
+        private readonly Stream _utf8Json;
+        private readonly JsonSerializerOptions? _options;
+
+        public AsyncEnumerableStreamingDeserializer(Stream utf8Json, JsonSerializerOptions? options)
+        {
+            _utf8Json = utf8Json;
+            _options = options;
+        }
+
+        public IAsyncEnumerator<TValue> GetAsyncEnumerator(CancellationToken cancellationToken)
+        {
+            return new Enumerator(_utf8Json, _options, cancellationToken);
+        }
+
+        private sealed class Enumerator : IAsyncEnumerator<TValue>
```

Comment:
Couldn't this be implemented with an async iterator? e.g. (GitHub-compiled)
```C#
public async IAsyncEnumerator<TValue> GetAsyncEnumerator(CancellationToken cancellationToken)
{
    using ReadAsyncState asyncState = new ReadAsyncState(typeof(Queue<TValue>), options, cancellationToken);
    bool isFinalBlock = false;
    while (!isFinalBlock)
    {
        isFinalBlock = await JsonSerializer.ReadFromStream(_utf8Json, _asyncState).ConfigureAwait(false);
        JsonSerializer.ContinueDeserialize<Queue<TValue>>(asyncState, isFinalBlock);
        if (_asyncState.ReadStack.Current.ReturnValue is Queue<TValue?> queue)
        {
            while (queue.TryDequeue(out TValue value)
            {
                yield return value;
            }
        }        
    }
}
```

Reply:
Good point, thanks.